### PR TITLE
[ROCm] Cleanup ir_emitter atomic support for ROCM

### PR DESCRIPTION
**Pull request by draganmladjenovic**
Use a single instruction fp atomic intrinsic when possible. Support FP64 and BF16. Match atomic ordering to that of atomic* functions from HIP header.

Avoid usage of inttoptr in EmitAtomicOperationUsingCAS This alows the backend to select global atomic instruction and also enables SROA on cas_new_output.

### CODE REVIEW

File: xla/service/gpu/tests/gpu_atomic_test.cc [hunk 1]
```diff
@@ -100,24 +108,42 @@ TEST_F(GpuAtomicTest, TestAddAtomicF32) {
...
-CHECK: atomicrmw fadd ptr addrspace(1) %[[ADDR:.*]], float %[[VALUE:.*]] syncscope("agent") seq_cst
+  HloModuleConfig config;
+  DebugOptions debug_options = config.debug_options();
+  debug_options.set_xla_gpu_ftz(true);
+  config.set_debug_options(debug_options);
+
+  TF_ASSERT_OK_AND_ASSIGN(auto module,
+                          ParseAndReturnVerifiedModule(hlo_string, config));
+
+  CompileAndVerifyIr(std::move(module), is_built_with_rocm_ ? R"(
```
Comment (jreiffers) [inline xla/service/gpu/tests/gpu_atomic_test.cc:119]: Nit: I think this is a bit more readable as

```
if (is_build_with_rocm_) {
  CompileAndVerifyIr
} else {
  CompileAndVerifyIr
}
```

(Assuming there's some reason why the separate test suite doesn't work).

File: xla/service/gpu/tests/gpu_atomic_test.cc [hunk 2]
```diff
@@ -78,6 +78,14 @@ CHECK-NOT: store atomic{{.*}}unordered, align 4
 }
 
 TEST_F(GpuAtomicTest, TestAddAtomicF32) {
+  if (is_built_with_rocm_ &&
```
Comment (jreiffers) [inline xla/service/gpu/tests/gpu_atomic_test.cc:81]: Let's create a separate test for AMD. I believe you should be able to use `GTEST_SKIP` in `SetUp` to prevent execution on Nvidia hardware then.

File: xla/service/gpu/tests/gpu_atomic_test.cc [hunk 3]
```diff
@@ -100,24 +108,42 @@ TEST_F(GpuAtomicTest, TestAddAtomicF32) {
     }
 )";
 
-  CompileAndVerifyIr(hlo_string, is_built_with_rocm_ ? R"(
-CHECK: atomicrmw fadd ptr addrspace(1) %[[ADDR:.*]], float %[[VALUE:.*]] syncscope("agent") seq_cst
+  HloModuleConfig config;
```
Comment (jreiffers) [inline xla/service/gpu/tests/gpu_atomic_test.cc:111]: Can you add some comments to explain the purpose of this?

File: xla/service/gpu/tests/gpu_atomic_test.cc [hunk 4]
```diff
@@ -100,24 +108,42 @@ TEST_F(GpuAtomicTest, TestAddAtomicF32) {
...
            .IsAtLeast(6)) {
     return;
   }
 
+  if (is_built_with_rocm_ &&
+      !backend()
+           .default_stream_executor()
+           ->GetDeviceDescription()
+           .rocm_compute_capability()
+           .gfx9_mi200_or_later()) {
```
Comment (jreiffers) [inline xla/service/gpu/tests/gpu_atomic_test.cc:143]: `has_fp64_atomics_support`

File: xla/service/gpu/tests/gpu_atomic_test.cc [hunk 5]
```diff
@@ -78,6 +78,14 @@ CHECK-NOT: store atomic{{.*}}unordered, align 4
 }
 
 TEST_F(GpuAtomicTest, TestAddAtomicF32) {
+  if (is_built_with_rocm_ &&
+      !backend()
+           .default_stream_executor()
+           ->GetDeviceDescription()
+           .rocm_compute_capability()
+           .gfx9_mi100_or_later()) {
```
Comment (jreiffers) [inline xla/service/gpu/tests/gpu_atomic_test.cc:86]: `has_fp32_atomics_support`

**Comment by jreiffers** [comment]
Thanks for your contribution. Just a heads up: this code path is essentially deprecated, so I'd recommend not spending any more effort on it. Investing in making the MLIR emitters work for AMD is definitely the better choice at this point.

File: xla/service/gpu/ir_emitter_nested.cc [hunk 6]
```diff
@@ -306,12 +307,23 @@ llvm::Value* AddrCastToDefault(llvm::Value* arg, llvm::IRBuilder<>& b) {
   return arg;
 }
 
-void EmitAMDGPUAtomicAdd(llvm::IRBuilder<>* builder,
-                         llvm::Value* output_address, llvm::Value* source) {
+void EmitAMDGPUAtomicFAdd(llvm::IRBuilder<>* builder,
+                          IrEmitterContext& ir_emitter_context,
+                          llvm::Value* output_address, llvm::Value* source,
+                          PrimitiveType element_type) {
+  static bool no_paired_atomics = []() -> bool {
```
Comment (jreiffers) [inline xla/service/gpu/ir_emitter_nested.cc:314]: Let's name this `paired_atomics` and remove the negation where it's used.

File: xla/service/gpu/ir_emitter_nested.cc [hunk 7]
```diff
@@ -306,12 +307,23 @@ llvm::Value* AddrCastToDefault(llvm::Value* arg, llvm::IRBuilder<>& b) {
...
+    const char* flag = std::getenv("ROCM_NO_PAIRED_ATOMICS");
```
Comment (jreiffers) [inline xla/service/gpu/ir_emitter_nested.cc:315]: I'm guessing this should be in `XLA_FLAGS`. If for some reason it can't, let's at least shorten this to `return flag && ...`.

File: xla/service/gpu/ir_emitter_nested.cc [hunk 8]
```diff
@@ -324,15 +336,63 @@ void EmitAMDGPUAtomicAdd(llvm::IRBuilder<>* builder,
...
+    auto alignment =
+        builder->CreateAnd(intptr, llvm::ConstantInt::get(i64, 2ull));
+    intptr = builder->CreateAnd(intptr, llvm::ConstantInt::get(i64, ~3ull));
+    output_ptr = builder->CreateIntToPtr(intptr, half2ptr);
+
+    auto shift = builder->CreateShl(builder->CreateTrunc(alignment, i32), 3);
+    auto i16src = builder->CreateBitCast(source, i16);
+    auto intsrc = builder->CreateZExt(i16src, i32);
+    source = builder->CreateShl(intsrc, shift);
+    source = builder->CreateBitCast(source, half2type);
```
Comment (jreiffers) [inline xla/service/gpu/ir_emitter_nested.cc:372]: The logic above here can probably use a comment or two.

File: xla/service/gpu/ir_emitter_nested.cc [hunk 9]
```diff
@@ -324,15 +336,63 @@ void EmitAMDGPUAtomicAdd(llvm::IRBuilder<>* builder,
...
+      llvm::AtomicOrdering::Monotonic,
       builder->getContext().getOrInsertSyncScopeID("agent"));
 }
 
-llvm::SyncScope::ID DetermineSyncScope(llvm::Module* module) {
-  return IsAMDGPU(module) ? module->getContext().getOrInsertSyncScopeID("agent")
-                          : llvm::SyncScope::System;
+std::tuple<llvm::SyncScope::ID, llvm::AtomicOrdering>
+DetermineSyncScopeAndOrdering(llvm::Module* module) {
+  return IsAMDGPU(module)
```
Comment (jreiffers) [inline xla/service/gpu/ir_emitter_nested.cc:390]: Probably more readable as an `if`.

**Review by jreiffers — CHANGES_REQUESTED** [review]

**Comment by dimitar-asenov** [comment]
@draganmladjenovic What is the state of this PR, do you still intend to merge it? Does the work on MLIR emitters obviate this PR? If you no longer intend to merge this PR (in favor of other work), please mark it as draft if you still need it for reference or close it altogether if you don't need it.

**Comment by draganmladjenovic** [comment]
I plan to merge it. I still would make life easier for our downstream. Probably some time next week.

**Comment by dimitar-asenov** [comment]
Hi @draganmladjenovic, could you please provide an update? Please consider making this PR a draft, in case you won't work on it in the near future. Thank you.

**Comment by dimitar-asenov** [comment]
A lot of the code that this PR touches has been deleted. I'm closing it. If you need these changes implemented, please target the new MLIR-based emitters.
